Add migration script for G Suite refactor

adrienthebo opened the pull request:
This pull request adds a first pass of a migration script that can detect project-factory resources in a state file and intelligently rename them.

aaron-lane commented on test/helpers/test_migrate.py:
```diff
@@ -0,0 +1,149 @@
+
+class TestGSuiteMigration(unittest.TestCase):
+    def setUp(self):
+        self.resources = [
```
`from_path` should have tests to prove its correctness, or `self.resources` should be hardcoded with the expected values so it does not depend on the correctness of `from_path`.

adrienthebo replied:
Added test coverage for `from_path` and some light input validation (though if `terraform state list` outputs bad values, something has gone horribly wrong)

aaron-lane commented on test/helpers/test_migrate.py:
```diff
@@ -0,0 +1,149 @@
+
+class TestTerraformModule(unittest.TestCase):
+    def setUp(self):
+        self.resources = [
```
Ditto regarding `from_path`.

adrienthebo commented on test/helpers/test_migrate.py:
```diff
@@ -0,0 +1,149 @@
+        self.resources = [
+            migrate.TerraformResource.from_path(path)
+            for path in TERRAFORM_STATE_LIST.split("\n")
+            if path.startswith("module.project-factory")
```
I think that I intended to check that we didn't touch any other resources but it appears that I didn't write tests for that. Think it's worth the effort to write those tests or should we just rip out the extra resources and the `.startswith` filter?

adrienthebo replied:
Addressed in https://github.com/terraform-google-modules/terraform-google-project-factory/pull/35/commits/50b5f827873b9cc33e283709b3b397c2e8143f0f

adrienthebo commented on test/helpers/test_migrate.py:
```diff
@@ -0,0 +1,149 @@
+
+import migrate  # noqa: E402
+
+TERRAFORM_STATE_LIST = """google_project_iam_member.additive_sa_role
```
Addressed in https://github.com/terraform-google-modules/terraform-google-project-factory/pull/35/commits/319f3012ba0406a1cf9e56b7c5b6545e28ece28a

aaron-lane commented on helpers/migrate.py:
```diff
@@ -0,0 +1,292 @@
+    """
+
+    CORE_PROJECT_FACTORY_SELECTORS = [
+        ("google_organization", "org"),
```
`google_organization.org`  was extracted to the `google_organization` submodule as the logic around it was duplicated in `gsuite_enabled` and `core_project_factory`.

adrienthebo replied:
I've dropped this from the refactored resource list, and since it's a data source it won't cause issues.

aaron-lane commented on helpers/migrate.py:
```diff
@@ -0,0 +1,292 @@
+    CORE_PROJECT_FACTORY_SELECTORS = [
+        ("google_organization", "org"),
+        ("google_compute_default_service_account", "default"),
+        ("null_data_source", "data_final_group_email"),
```
`data_final_group_email` was refactored out of existence.

aaron-lane commented on helpers/migrate.py:
```diff
@@ -0,0 +1,292 @@
+        ("google_organization", "org"),
+        ("google_compute_default_service_account", "default"),
+        ("null_data_source", "data_final_group_email"),
+        ("null_data_source", "data_group_email_format"),
```
`data_group_email_format` was refactored out of existence.

adrienthebo replied:
I've dropped this from the refactored resource list, and since it's a data source it won't cause issues.

aaron-lane commented on helpers/migrate.py:
```diff
@@ -0,0 +1,292 @@
+        ("null_data_source", "data_final_group_email"),
+        ("null_data_source", "data_group_email_format"),
+        ("google_compute_shared_vpc_service_project", None),
+        ("google_compute_subnetwork_iam_member", None),
```
`google_compute_subnetwork_iam_member.group_role_to_vpc_subnets` remains in `gsuite_enabled` so this rule must be restricted.

aaron-lane commented on helpers/migrate.py:
```diff
@@ -0,0 +1,292 @@
+        ("google_compute_shared_vpc_service_project", None),
+        ("google_compute_subnetwork_iam_member", None),
+        ("google_project", None),
+        ("google_project_iam_member", None),
```
`google_project_iam_member.gsuite_group_role` and `google_project_iam_member.controlling_group_vpc_membership` remain in `gsuite_enabled` so this rule must be restricted.

In case we test this script against test state, `google_project_iam_member.additive_sa_role` and `google_project_iam_member.additive_shared_vpc_role` remain in `full` fixture.

adrienthebo replied:
The refactoring of the shared VPC `roles/compute.networkUser` role means that we can't perform a simple migration through `terraform state mv`. I'm getting this diff:

```
An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  + create
-/+ destroy and then create replacement

Terraform will perform the following actions:

  + module.project-factory.google_project_iam_member.controlling_group_vpc_membership
      id:      <computed>
      etag:    <computed>
      member:  "group:pf-gsuite-migrate-group@phoogle.net"
      project: "thebo-host-c15e"
      role:    "roles/compute.networkUser"

-/+ module.project-factory.module.project-factory.google_project_iam_member.controlling_group_vpc_membership[1] (new resource required)
      id:      "thebo-host-c15e/roles/compute.networkUser/group:pf-gsuite-migrate-group@phoogle.net" => <computed> (forces new resource)
      etag:    "BwV872YZKPU=" => <computed>
      member:  "group:pf-gsuite-migrate-group@phoogle.net" => "serviceAccount:303471683246@cloudservices.gserviceaccount.com" (forces new resource)
      project: "thebo-host-c15e" => "thebo-host-c15e"
      role:    "roles/compute.networkUser" => "roles/compute.networkUser"

-/+ module.project-factory.module.project-factory.google_project_iam_member.controlling_group_vpc_membership[2] (new resource required)
      id:      "thebo-host-c15e/roles/compute.networkUser/serviceAccount:303471683246@cloudservices.gserviceaccount.com" => <computed> (forces new resource)
      etag:    "BwV872YZKPU=" => <computed>
      member:  "serviceAccount:303471683246@cloudservices.gserviceaccount.com" => "serviceAccount:project-service-account@pf-gsuite-migrate-group-23dc.iam.gserviceaccount.com" (forces new resource)
      project: "thebo-host-c15e" => "thebo-host-c15e"
      role:    "roles/compute.networkUser" => "roles/compute.networkUser"
```

I think we're stuck with re-creating these resources or doing some serious state surgery, which increases the risk of something going wrong.

aaron-lane replied:
Is there a risk in recreating those resources?

aaron-lane commented on helpers/migrate.py:
```diff
@@ -0,0 +1,292 @@
+        ("google_project_iam_member", None),
+        ("google_project_service", None),
+        ("google_project_usage_export_bucket", None),
+        ("google_service_account", None),
```
In case we test this script on test state, `google_service_account.extra_service_account` remains in `full` fixture.

adrienthebo replied:
These resources should only be moved when found within a project-factory Terraform module, so this won't be touched.

aaron-lane commented on helpers/migrate.py:
```diff
@@ -0,0 +1,292 @@
+        ("google_project_service", None),
+        ("google_project_usage_export_bucket", None),
+        ("google_service_account", None),
+        ("google_service_account_iam_member", None),
```
`google_service_account_iam_member.service_account_grant_to_group` remains in `gsuite_enabled` so this rule must be restricted.

In case we test this script on test state, `google_service_account_iam_member.additive_service_account_grant_to_group` remains in `full` fixture.

aaron-lane commented on helpers/migrate.py:
```diff
@@ -0,0 +1,292 @@
+        ("google_service_account", None),
+        ("google_service_account_iam_member", None),
+        ("google_storage_bucket", None),
+        ("google_storage_bucket_iam_member", None),
```
`google_storage_bucket_iam_member.group_storage_admin_on_project_bucket` remains in `gsuite_enabled` so this rule must be restricted.

adrienthebo commented:
I've refactored the migration module to better handle multiple modules, but I need to spend a lot of time handling the `gsuite_enabled` path. With the old module structure, the `gsuite_enabled` module was at the root of the module and so those resources didn't need refactoring, but the extraction of the `google_group` module broke that assumption and added new migration paths that we'll have to account for.

aaron-lane commented:
@adrienthebo I adjusted `aaron-lane/gsuite-refactor`; the `google_group` module wasn't actually necessary .

aaron-lane commented:
@adrienthebo Is there a risk in recreating those resources?

adrienthebo commented:
I've rebased this on top of `aaron-lane/gsuite-refactor`, so a reset against the updated branch may be required. In addition I've verified that the recreation of the IAM roles does not create an outage; the most I saw was a very brief increase in the instance ping time but no real interruption.

adrienthebo commented:
I've updated the tests and used the output of `terraform state list` from a test environment as fixtures. The tests feel a little contrived though, could you eyeball them and see if we're actually testing anything interesting when checking the computed moves?

aaron-lane commented:
@adrienthebo: I don't see any major issues with the tests. The stubbed state list of resources could be shortened for clarity; I don't think 17 counts of the same resources really provide any more value than say 2 counts.

adrienthebo commented:
@aaron-lane fair point - I've trimmed that down to 3 subnets, and added some additional tests to make sure that we don't touch resources in other modules. Care to give another look?
